File: src/main/kotlin/org/gameflow/pass/Pass.kt
package org.gameflow.pass

import org.gameflow.entity.Entity
import org.gameflow.state.GameState
import org.gameflow.Game
import org.gameflow.entity.EntityGroup
import java.util.Collection
import org.gameflow.entity.EntityGroups


/**
 * Something that processes all entities matching specified criteria each game loop, or at some other intervals.
 */
trait Pass {

    fun init(game: Game)
    fun deInit(game: Game)

    fun startPass(game: Game)
    fun apply(entity: Entity, game: Game)
    fun endPass(game: Game)

}

File: src/main/scala/org/gameflow/core/pass/Pass.scala
package org.gameflow.core.pass

import org.gameflow.core.entity.Entity
import org.gameflow.core.{LifeCycled, Game}


/**
 * Something that processes all entities matching specified criteria each game loop, or at some other intervals.
 */
trait Pass extends LifeCycled with Comparable[Pass] {

  def startPass(game: Game) {}

  def handleEntity(entity: Entity, game: Game)

  def endPass(game: Game) {}

  def compareTo(o: Pass): Int = {
    val selfHash = hashCode()
    val otherHash = o.hashCode()
    if (selfHash < otherHash) -1
    else if (selfHash > otherHash) 1
    else 0
  }
}